File: verman-py/file_manager.py
"""
File scanning, hashing, backup, restore, and export helpers.
"""

import fnmatch
import hashlib
import os
import shutil
import tempfile
import time
from datetime import datetime
from typing import Dict, Iterable, List, Optional, Tuple

from logger import logger as app_logger
from models import (
    ACTIVE_FILE_STATUSES,
    DEFAULT_IGNORE_PATTERNS,
    BlockedFile,
    FileState,
)
from project_paths import get_backup_dir, get_ignore_file_path


class FileManager:
    """File operations for the managed workspace."""

    def __init__(self, workspace_path: str):
        self.workspace_path = os.path.abspath(workspace_path)
        self._hash_cache: Dict[str, Tuple[str, int, int, float]] = {}
        self._cache_max_size = 1000
        self._cache_ttl = 300.0

    def clear_hash_cache(self):
        self._hash_cache.clear()

    def scan_workspace(
        self,
        ignore_patterns: Optional[List[str]] = None,
        indexed_files: Optional[Dict[str, Dict[str, int]]] = None,
    ) -> Tuple[Dict[str, FileState], List[BlockedFile]]:
        """
        Scan the workspace and reuse stored hashes when size and mtime match.
        """
        indexed_files = indexed_files or {}
        current_files: Dict[str, FileState] = {}
        blocked_files: List[BlockedFile] = []

        for relative_path, file_path in self._iter_visible_files(ignore_patterns):
            try:
                stat_result = os.stat(file_path)
                file_size = stat_result.st_size
                mtime_ns = stat_result.st_mtime_ns
            except OSError as exc:
                app_logger.warning(f"跳过文件 {relative_path}: {exc}")
                continue

            indexed_state = indexed_files.get(relative_path)
            if (
                indexed_state
                and indexed_state.get("file_size") == file_size
                and indexed_state.get("mtime_ns") == mtime_ns
                and indexed_state.get("file_hash")
            ):
                file_hash = indexed_state["file_hash"]
            else:
                file_hash = self._calculate_file_hash(file_path, file_size, mtime_ns)

            if not file_hash:
                continue

            current_files[relative_path] = FileState(
                relative_path=relative_path,
                file_hash=file_hash,
                file_size=file_size,
                mtime_ns=mtime_ns,
            )

        return current_files, blocked_files

    def list_workspace_files(self, ignore_patterns: Optional[List[str]] = None) -> List[str]:
        return [
            relative_path
            for relative_path, _ in self._iter_visible_files(ignore_patterns)
        ]

    def read_relative_file(self, relative_path: str) -> bytes:
        full_path = os.path.join(self.workspace_path, relative_path)
        return self._read_file_content(full_path)

    def restore_files(
        self,
        version_files: List[Dict],
        ignore_patterns: Optional[List[str]] = None,
        backup_current: bool = True,
    ) -> Dict[str, object]:
        try:
            if backup_current:
                self._backup_current_state(ignore_patterns)

            desired_active = {
                file_info["relative_path"]: file_info
                for file_info in version_files
                if file_info["file_status"] in ACTIVE_FILE_STATUSES
            }
            desired_paths = set(desired_active.keys())
            current_paths = set(self.list_workspace_files(ignore_patterns))

            restored_count = 0
            removed_count = 0
            warnings: List[str] = []

            for extra_path in sorted(current_paths - desired_paths):
                full_path = os.path.join(self.workspace_path, extra_path)
                if os.path.exists(full_path):
                    os.remove(full_path)
                    removed_count += 1
                    self._cleanup_empty_directories(os.path.dirname(full_path))

            for file_data in version_files:
                relative_path = file_data["relative_path"]
                file_status = file_data["file_status"]
                target_path = os.path.join(self.workspace_path, relative_path)

                if file_status == "delete":
                    if os.path.exists(target_path):
                        os.remove(target_path)
                        removed_count += 1
                        self._cleanup_empty_directories(os.path.dirname(target_path))
                    continue

                file_content = file_data.get("file_content")
                if file_content is None:
                    raise ValueError(f"版本文件缺少内容: {relative_path}")

                target_dir = os.path.dirname(target_path)
                if target_dir:
                    os.makedirs(target_dir, exist_ok=True)
                with open(target_path, "wb") as file_handle:
                    file_handle.write(file_content)
                restored_count += 1

            return {
                "restored_count": restored_count,
                "removed_count": removed_count,
                "warnings": warnings,
            }
        except Exception as exc:
            app_logger.error(f"文件恢复失败: {exc}")
            raise

    def export_version_files(self, version_files: List[Dict], export_path: str) -> bool:
        try:
            for file_data in version_files:
                if file_data["file_status"] == "delete":
                    continue

                relative_path = file_data["relative_path"]
                file_content = file_data.get("file_content")
                if file_content is None:
                    raise ValueError(f"导出文件缺少内容: {relative_path}")

                target_path = os.path.join(export_path, relative_path)
                target_dir = os.path.dirname(target_path)
                if target_dir:
                    os.makedirs(target_dir, exist_ok=True)
                with open(target_path, "wb") as file_handle:
                    file_handle.write(file_content)

            return True
        except Exception as exc:
            app_logger.error(f"导出失败: {exc}")
            return False

    def _iter_visible_files(self, ignore_patterns: Optional[List[str]]) -> Iterable[Tuple[str, str]]:
        all_ignore_patterns = list(DEFAULT_IGNORE_PATTERNS)
        all_ignore_patterns.extend(self._load_ignore_file())
        all_ignore_patterns.extend(ignore_patterns or [])

        if not os.path.exists(self.workspace_path):
            raise FileNotFoundError(f"工作区路径不存在: {self.workspace_path}")
        if not os.access(self.workspace_path, os.R_OK):
            raise PermissionError(f"工作区路径不可读: {self.workspace_path}")

        max_files = 10000
        yielded = 0

        for root, dirs, files in os.walk(self.workspace_path):
            relative_root = os.path.relpath(root, self.workspace_path)
            if relative_root.startswith(".."):
                continue

            dirs[:] = [
                dir_name
                for dir_name in dirs
                if not self._should_ignore(
                    os.path.relpath(os.path.join(root, dir_name), self.workspace_path),
                    all_ignore_patterns,
                    is_dir=True,
                )
            ]

            for file_name in files:
                file_path = os.path.join(root, file_name)
                file_path = self._resolve_symlink_path(file_path)
                if file_path is None:
                    continue

                relative_path = os.path.relpath(file_path, self.workspace_path)
                if (
                    relative_path.startswith("..")
                    or ".." in relative_path.split(os.sep)
                    or os.path.isabs(relative_path)
                ):
                    continue

                if self._should_ignore(relative_path, all_ignore_patterns, is_dir=False):
                    continue

                if not os.access(file_path, os.R_OK):
                    continue

                yielded += 1
                if yielded > max_files:
                    app_logger.warning(f"文件数量超过限制 ({max_files})，停止扫描")
                    return

                yield relative_path.replace("\\", "/"), file_path

    def _resolve_symlink_path(self, file_path: str) -> Optional[str]:
        if not os.path.islink(file_path):
            return file_path

        try:
            link_target = os.path.realpath(file_path)
            target_relative = os.path.relpath(link_target, self.workspace_path)
            if target_relative.startswith(".."):
                app_logger.debug(f"跳过外部链接 {file_path} -> {link_target}")
                return None
            return link_target
        except (OSError, ValueError):
            app_logger.debug(f"跳过无效链接 {file_path}")
            return None

    def _calculate_file_hash(self, file_path: str, file_size: int, mtime_ns: int) -> str:
        relative_path = os.path.relpath(file_path, self.workspace_path).replace("\\", "/")
        cached_result = self._get_cached_hash(relative_path, file_size, mtime_ns)
        if cached_result is not None:
            return cached_result

        try:
            hash_md5 = hashlib.md5()
            with open(file_path, "rb") as file_handle:
                chunk_size = 4096 if file_size < 10 * 1024 * 1024 else 64 * 1024
                for chunk in iter(lambda: file_handle.read(chunk_size), b""):
                    hash_md5.update(chunk)

            calculated_hash = hash_md5.hexdigest()
            self._update_hash_cache(relative_path, calculated_hash, file_size, mtime_ns)
            return calculated_hash
        except (IOError, OSError) as exc:
            app_logger.warning(f"计算文件哈希失败 {file_path}: {exc}")
            return ""

    def _get_cached_hash(
        self, relative_path: str, file_size: int, mtime_ns: int
    ) -> Optional[str]:
        cached_entry = self._hash_cache.get(relative_path)
        if not cached_entry:
            return None

        cached_hash, cached_size, cached_mtime_ns, cached_at = cached_entry
        if time.time() - cached_at > self._cache_ttl:
            self._hash_cache.pop(relative_path, None)
            return None

        if cached_size != file_size or cached_mtime_ns != mtime_ns:
            self._hash_cache.pop(relative_path, None)
            return None

        return cached_hash

    def _update_hash_cache(self, relative_path: str, file_hash: str, file_size: int, mtime_ns: int):
        if len(self._hash_cache) >= self._cache_max_size:
            self._cleanup_hash_cache()

        self._hash_cache[relative_path] = (file_hash, file_size, mtime_ns, time.time())

    def _cleanup_hash_cache(self):
        if not self._hash_cache:
            return

        sorted_items = sorted(self._hash_cache.items(), key=lambda item: item[1][3])
        keep_count = int(len(sorted_items) * 0.75)
        self._hash_cache = dict(sorted_items[keep_count:])

    def _read_file_content(self, file_path: str) -> bytes:
        with open(file_path, "rb") as file_handle:
            return file_handle.read()

    def _load_ignore_file(self) -> List[str]:
        ignore_file_path = get_ignore_file_path(self.workspace_path)
        patterns = []
        try:
            if os.path.exists(ignore_file_path):
                with open(ignore_file_path, "r", encoding="utf-8") as file_handle:
                    for line in file_handle:
                        line = line.strip()
                        if line and not line.startswith("#"):
                            patterns.append(line)
        except (IOError, OSError, UnicodeDecodeError) as exc:
            app_logger.warning(f"读取忽略文件失败: {exc}")
        return patterns

    def _should_ignore(self, name: str, ignore_patterns: List[str], is_dir: bool = False) -> bool:
        normalized_name = name.replace("\\", "/").strip()
        if normalized_name.startswith("./"):
            normalized_name = normalized_name[2:]
        normalized_name = normalized_name.strip("/")
        basename = os.path.basename(normalized_name.rstrip("/"))

        for pattern in ignore_patterns:
            normalized_pattern = pattern.replace("\\", "/").strip()
            if not normalized_pattern:
                continue
            if normalized_pattern.startswith("./"):
                normalized_pattern = normalized_pattern[2:]

            if normalized_pattern.endswith("/"):
                if not is_dir:
                    continue
                dir_pattern = normalized_pattern.rstrip("/")
                if (
                    normalized_name == dir_pattern
                    or normalized_name.startswith(dir_pattern + "/")
                    or basename == dir_pattern
                    or fnmatch.fnmatch(normalized_name, dir_pattern)
                    or fnmatch.fnmatch(basename, dir_pattern)
                ):
                    return True
                continue

            if (
                fnmatch.fnmatch(basename, normalized_pattern)
                or fnmatch.fnmatch(normalized_name, normalized_pattern)
            ):
                return True

        return False

    def _backup_current_state(self, ignore_patterns: Optional[List[str]] = None):
        backup_dir = get_backup_dir(self.workspace_path)
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        backup_path = os.path.join(backup_dir, f"backup_{timestamp}")

        os.makedirs(backup_path, exist_ok=True)
        backed_up_count = 0

        for relative_path in self.list_workspace_files(ignore_patterns):
            source_path = os.path.join(self.workspace_path, relative_path)
            target_path = os.path.join(backup_path, relative_path)
            target_dir = os.path.dirname(target_path)
            if target_dir:
                os.makedirs(target_dir, exist_ok=True)
            shutil.copy2(source_path, target_path)
            backed_up_count += 1

        app_logger.info(f"当前状态已备份到: {backup_path}")
        app_logger.info(f"共备份 {backed_up_count} 个文件")

    def _cleanup_empty_directories(self, start_dir: str):
        current_dir = start_dir
        workspace_root = os.path.abspath(self.workspace_path)

        while current_dir and os.path.abspath(current_dir).startswith(workspace_root):
            if os.path.abspath(current_dir) == workspace_root:
                break
            try:
                os.rmdir(current_dir)
            except OSError:
                break
            current_dir = os.path.dirname(current_dir)

File: verman-rust/src-tauri/src/file_manager.rs
use std::collections::HashMap;
use std::io::Read;
use std::num::NonZeroUsize;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use lru::LruCache;
use rayon::prelude::*;
use serde::{Deserialize, Serialize};
use walkdir::WalkDir;

use md5::{Md5, Digest};

use crate::models::{
    ACTIVE_FILE_STATUSES, DEFAULT_IGNORE_PATTERNS, BlockedFile, FileState, VersionFileInfo,
    WorkspaceIndexEntry,
};

pub struct FileManager {
    workspace_path: PathBuf,
    hash_cache: LruCache<String, HashEntry>,
}

#[derive(Clone, Serialize, Deserialize)]
struct HashEntry {
    file_hash: String,
    file_size: i64,
    mtime_ns: i64,
    cached_at: u64,
}

impl FileManager {
    pub fn new(workspace_path: &Path) -> Self {
        let mut mgr = Self {
            workspace_path: workspace_path.to_owned(),
            hash_cache: LruCache::new(NonZeroUsize::new(1000).unwrap()),
        };
        mgr.load_hash_cache();
        mgr
    }

    pub fn save_hash_cache(&self) {
        let entries: Vec<(String, HashEntry)> = self.hash_cache.iter().map(|(k, v)| (k.clone(), v.clone())).collect();
        let path = crate::project_paths::get_hash_cache_path(&self.workspace_path);
        if entries.is_empty() {
            let _ = std::fs::remove_file(&path);
            return;
        }
        if let Ok(json) = serde_json::to_string(&entries) {
            let _ = std::fs::write(&path, &json);
        }
    }

    fn load_hash_cache(&mut self) {
        let path = crate::project_paths::get_hash_cache_path(&self.workspace_path);
        let json = match std::fs::read_to_string(&path) {
            Ok(s) => s,
            Err(_) => return,
        };
        let entries: Vec<(String, HashEntry)> = match serde_json::from_str(&json) {
            Ok(e) => e,
            Err(_) => return,
        };
        for (key, entry) in entries {
            self.hash_cache.put(key, entry);
        }
    }

    pub fn clear_hash_cache(&mut self) {
        self.hash_cache.clear();
    }

    pub fn scan_workspace(
        &mut self,
        ignore_patterns: &[String],
        indexed_files: &HashMap<String, WorkspaceIndexEntry>,
    ) -> (HashMap<String, FileState>, Vec<BlockedFile>) {
        let mut current_files: HashMap<String, FileState> = HashMap::new();
        let blocked_files: Vec<BlockedFile> = Vec::new();
        let mut pending_hash: Vec<(String, PathBuf)> = Vec::new();
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);

        for (relative_path, file_path) in self.iter_visible_files(ignore_patterns) {
            let metadata = match std::fs::metadata(&file_path) {
                Ok(m) => m,
                Err(e) => {
                    tracing::warn!(error = %e, path = %relative_path, "Skipping file");
                    continue;
                }
            };

            let file_size = metadata.len() as i64;
            let mtime_ns = metadata
                .modified()
                .ok()
                .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                .map(|d| d.as_nanos() as i64)
                .unwrap_or(0);

            // Try workspace index cache first
            let cached_hash = indexed_files.get(&relative_path).and_then(|indexed| {
                if indexed.file_size == file_size && indexed.mtime_ns == mtime_ns
                    && !indexed.file_hash.is_empty()
                {
                    Some(indexed.file_hash.clone())
                } else {
                    None
                }
            });

            // Try in-memory LRU cache
            let cached_hash = cached_hash.or_else(|| {
                self.hash_cache.get(&relative_path).and_then(|entry| {
                    if entry.file_size == file_size && entry.mtime_ns == mtime_ns
                        && now - entry.cached_at < 300
                    {
                        Some(entry.file_hash.clone())
                    } else {
                        None
                    }
                })
            });

            if let Some(hash) = cached_hash {
                current_files.insert(
                    relative_path.clone(),
                    FileState {
                        relative_path,
                        file_hash: hash,
                        file_size,
                        mtime_ns,
                    },
                );
            } else {
                pending_hash.push((relative_path, file_path));
            }
        }

        // Parallel phase: compute hashes for cache misses
        let hash_results: Vec<(String, String)> = pending_hash
            .par_iter()
            .map(|(rel_path, full_path)| {
                let hash = Self::compute_md5(full_path);
                (rel_path.clone(), hash)
            })
            .filter(|(_, hash)| !hash.is_empty())
            .collect();

        // Sequential: update cache and results
        for (relative_path, file_hash) in &hash_results {
            if file_hash.is_empty() {
                continue;
            }
            // Re-check metadata (may have changed since first pass)
            let full_path = self.workspace_path.join(relative_path);
            let metadata = std::fs::metadata(&full_path).ok();
            let file_size = metadata.as_ref().map(|m| m.len() as i64).unwrap_or(0);
            let mtime_ns = metadata
                .and_then(|m| m.modified().ok())
                .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
                .map(|d| d.as_nanos() as i64)
                .unwrap_or(0);

            current_files.insert(
                relative_path.clone(),
                FileState {
                    relative_path: relative_path.clone(),
                    file_hash: file_hash.clone(),
                    file_size,
                    mtime_ns,
                },
            );
            self.hash_cache.put(
                relative_path.clone(),
                HashEntry {
                    file_hash: file_hash.clone(),
                    file_size,
                    mtime_ns,
                    cached_at: now,
                },
            );
        }

        (current_files, blocked_files)
    }

    pub fn list_workspace_files(&self, ignore_patterns: &[String]) -> Vec<String> {
        self.iter_visible_files(ignore_patterns)
            .into_iter()
            .map(|(path, _)| path)
            .collect()
    }

    pub fn workspace_path(&self) -> &Path {
        &self.workspace_path
    }

    pub fn read_relative_file(&self, relative_path: &str) -> Result<Vec<u8>, String> {
        let full_path = self.workspace_path.join(relative_path);
        std::fs::read(&full_path).map_err(|e| format!("Failed to read {}: {}", relative_path, e))
    }

    pub fn restore_files(
        &self,
        version_files: &[VersionFileInfo],
        ignore_patterns: &[String],
        backup_current: bool,
    ) -> Result<(i64, i64, Vec<String>), String> {
        if backup_current {
            self.backup_current_state(ignore_patterns)?;
        }

        let desired_active: HashMap<&str, &VersionFileInfo> = version_files
            .iter()
            .filter(|f| ACTIVE_FILE_STATUSES.contains(&f.file_status.as_str()))
            .map(|f| (f.relative_path.as_str(), f))
            .collect();

        let desired_paths: std::collections::HashSet<&&str> =
            desired_active.keys().collect();
        let current_paths: std::collections::HashSet<String> =
            self.list_workspace_files(ignore_patterns).into_iter().collect();

        let mut removed_count: i64 = 0;
        let mut restored_count: i64 = 0;
        let warnings: Vec<String> = Vec::new();

        // Remove files not in target version
        for extra_path in current_paths.iter() {
            if !desired_paths.contains(&extra_path.as_str()) {
                let full_path = self.workspace_path.join(extra_path);
                if full_path.exists() {
                    std::fs::remove_file(&full_path)
                        .map_err(|e| format!("Failed to remove {}: {}", extra_path, e))?;
                    removed_count += 1;
                    self.cleanup_empty_directories(full_path.parent().unwrap());
                }
            }
        }

        // Restore files from version
        for file in version_files {
            let target_path = self.workspace_path.join(&file.relative_path);
            match file.file_status.as_str() {
                "delete" => {
                    if target_path.exists() {
                        std::fs::remove_file(&target_path)
                            .map_err(|e| format!("Failed to remove {}: {}", file.relative_path, e))?;
                        removed_count += 1;
                        self.cleanup_empty_directories(target_path.parent().unwrap());
                    }
                }
                "add" | "modify" | "unmodified" => {
                    let content = match file.file_content.as_ref() {
                        Some(bytes) => bytes.clone(),
                        None => {
                            // Try blob storage
                            let blob_path = crate::project_paths::get_blob_path(
                                &self.workspace_path, &file.file_hash
                            );
                            std::fs::read(&blob_path).map_err(|e| {
                                format!("Missing content and blob for {}: {}", file.relative_path, e)
                            })?
                        }
                    };
                    if let Some(parent) = target_path.parent() {
                        std::fs::create_dir_all(parent)
                            .map_err(|e| format!("Failed to create dir {:?}: {}", parent, e))?;
                    }
                    std::fs::write(&target_path, content)
                        .map_err(|e| format!("Failed to write {}: {}", file.relative_path, e))?;
                    restored_count += 1;
                }
                _ => {}
            }
        }

        Ok((restored_count, removed_count, warnings))
    }

    pub fn export_version_files(
        &self,
        version_files: &[VersionFileInfo],
        export_path: &Path,
    ) -> Result<bool, String> {
        for file in version_files {
            if file.file_status == "delete" {
                continue;
            }
            let content = match file.file_content.as_ref() {
                Some(bytes) => bytes.clone(),
                None => {
                    let blob_path = crate::project_paths::get_blob_path(
                        &self.workspace_path, &file.file_hash
                    );
                    std::fs::read(&blob_path).map_err(|e| {
                        format!("Missing content and blob for {}: {}", file.relative_path, e)
                    })?
                }
            };
            let target_path = export_path.join(&file.relative_path);
            if let Some(parent) = target_path.parent() {
                std::fs::create_dir_all(parent)
                    .map_err(|e| format!("Failed to create dir {:?}: {}", parent, e))?;
            }
            std::fs::write(&target_path, content)
                .map_err(|e| format!("Failed to write {}: {}", file.relative_path, e))?;
        }
        Ok(true)
    }

    fn iter_visible_files(
        &self,
        ignore_patterns: &[String],
    ) -> Vec<(String, PathBuf)> {
        let all_ignore = self.build_ignore_patterns(ignore_patterns);
        let mut results = Vec::new();

        let walker = WalkDir::new(&self.workspace_path)
            .into_iter()
            .filter_entry(|entry| {
                let relative = entry
                    .path()
                    .strip_prefix(&self.workspace_path)
                    .unwrap_or(entry.path());
                let relative_str = relative.to_string_lossy().replace('\\', "/");
                if entry.file_type().is_dir() {
                    let name = relative_str.trim_end_matches('/').to_string();
                    !self.should_ignore(&name, &all_ignore, true)
                } else {
                    true
                }
            });

        for entry in walker {
            let entry = match entry {
                Ok(e) => e,
                Err(_) => continue,
            };
            if !entry.file_type().is_file() {
                continue;
            }

            let file_path = entry.path().to_owned();
            let relative = file_path
                .strip_prefix(&self.workspace_path)
                .unwrap_or(&file_path);
            let relative_str = relative.to_string_lossy().replace('\\', "/");

            if relative_str.starts_with("..") || relative_str.contains("../") {
                continue;
            }

            if self.should_ignore(&relative_str, &all_ignore, false) {
                continue;
            }

            if results.len() >= 10_000 {
                tracing::warn!("File count exceeded limit (10000), stopping scan");
                break;
            }

            results.push((relative_str, file_path));
        }

        results
    }

    fn build_ignore_patterns(&self, extra_patterns: &[String]) -> Vec<String> {
        let mut patterns: Vec<String> = DEFAULT_IGNORE_PATTERNS
            .iter()
            .map(|s| s.to_string())
            .collect();

        patterns.extend(self.load_ignore_file());
        patterns.extend(extra_patterns.iter().cloned());
        patterns
    }

    fn load_ignore_file(&self) -> Vec<String> {
        let ignore_path = self.workspace_path.join(".vermanignore");
        let content = match std::fs::read_to_string(&ignore_path) {
            Ok(c) => c,
            Err(_) => return Vec::new(),
        };

        content
            .lines()
            .map(|line| line.trim().to_string())
            .filter(|line| !line.is_empty() && !line.starts_with('#'))
            .collect()
    }

    fn should_ignore(&self, name: &str, ignore_patterns: &[String], is_dir: bool) -> bool {
        let normalized = name.replace('\\', "/").trim().to_string();
        let normalized = normalized.strip_prefix("./").unwrap_or(&normalized).to_string();
        let normalized = normalized.trim_end_matches('/').to_string();
        let basename = Path::new(&normalized)
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default();

        for pattern in ignore_patterns {
            let p = pattern.replace('\\', "/").trim().to_string();
            let p = p.strip_prefix("./").unwrap_or(&p).to_string();
            if p.is_empty() {
                continue;
            }

            if p.ends_with('/') {
                if !is_dir {
                    continue;
                }
                let dir_pattern = p.trim_end_matches('/');
                if normalized == dir_pattern
                    || normalized.starts_with(&format!("{}/", dir_pattern))
                    || basename == dir_pattern
                    || Self::fnmatch(&normalized, dir_pattern)
                    || Self::fnmatch(&basename, dir_pattern)
                {
                    return true;
                }
                continue;
            }

            if Self::fnmatch(&basename, &p) || Self::fnmatch(&normalized, &p) {
                return true;
            }
        }

        false
    }

    fn fnmatch(name: &str, pattern: &str) -> bool {
        if pattern.is_empty() {
            return name.is_empty();
        }

        let pattern_chars: Vec<char> = pattern.chars().collect();
        let name_chars: Vec<char> = name.chars().collect();

        Self::fnmatch_recursive(&name_chars, &pattern_chars, 0, 0)
    }

    fn fnmatch_recursive(
        name: &[char],
        pattern: &[char],
        ni: usize,
        pi: usize,
    ) -> bool {
        if pi == pattern.len() {
            return ni == name.len();
        }

        match pattern[pi] {
            '*' => {
                // Try matching zero or more characters
                if Self::fnmatch_recursive(name, pattern, ni, pi + 1) {
                    return true;
                }
                if ni < name.len()
                    && Self::fnmatch_recursive(name, pattern, ni + 1, pi)
                {
                    return true;
                }
                false
            }
            '?' => {
                if ni < name.len() {
                    Self::fnmatch_recursive(name, pattern, ni + 1, pi + 1)
                } else {
                    false
                }
            }
            '[' => {
                // Find the closing bracket
                let mut negated = false;
                if pi + 1 < pattern.len() && pattern[pi + 1] == '!' {
                    negated = true;
                }
                let mut matched = false;
                let close = pattern[pi..].iter().position(|&c| c == ']');
                if let Some(end) = close {
                    let end = pi + end;
                    let charset = &pattern[pi + 1..end];
                    // Simple bracket expression (no ranges for now)
                    if ni < name.len() {
                        matched = charset.contains(&name[ni]);
                        if negated {
                            matched = !matched;
                        }
                    }
                    if matched {
                        return Self::fnmatch_recursive(name, pattern, ni + 1, end + 1);
                    }
                    false
                } else {
                    // No closing bracket, treat as literal
                    if ni < name.len() && name[ni] == pattern[pi] {
                        Self::fnmatch_recursive(name, pattern, ni + 1, pi + 1)
                    } else {
                        false
                    }
                }
            }
            c => {
                if ni < name.len() && (name[ni] == c || (c == '/' && name[ni] == '\\')) {
                    Self::fnmatch_recursive(name, pattern, ni + 1, pi + 1)
                } else {
                    false
                }
            }
        }
    }


    fn compute_md5(file_path: &Path) -> String {
        let file = match std::fs::File::open(file_path) {
            Ok(f) => f,
            Err(e) => {
                tracing::warn!(error = %e, path = %file_path.display(), "Cannot open for hashing");
                return String::new();
            }
        };

        let metadata = match file.metadata() {
            Ok(m) => m,
            Err(_) => return String::new(),
        };
        let file_size = metadata.len();

        let chunk_size: usize = if file_size < 10 * 1024 * 1024 {
            4096
        } else {
            64 * 1024
        };

        let mut hasher = Md5::new();
        let mut reader = std::io::BufReader::with_capacity(chunk_size, file);
        let mut buffer = vec![0u8; chunk_size];

        loop {
            match reader.read(&mut buffer) {
                Ok(0) => break,
                Ok(n) => {
                    hasher.update(&buffer[..n]);
                }
                Err(e) => {
                    tracing::warn!(error = %e, path = %file_path.display(), "Hash read error");
                    return String::new();
                }
            }
        }

        format!("{:x}", hasher.finalize())
    }

    fn backup_current_state(&self, ignore_patterns: &[String]) -> Result<(), String> {
        let backup_dir = crate::project_paths::get_backup_dir(&self.workspace_path);
        let timestamp = chrono::Local::now().format("%Y%m%d_%H%M%S");
        let backup_path = backup_dir.join(format!("backup_{}", timestamp));

        std::fs::create_dir_all(&backup_path)
            .map_err(|e| format!("Failed to create backup dir: {}", e))?;

        let mut count = 0;
        for relative_path in self.list_workspace_files(ignore_patterns) {
            let source = self.workspace_path.join(&relative_path);
            let target = backup_path.join(&relative_path);
            if let Some(parent) = target.parent() {
                std::fs::create_dir_all(parent).ok();
            }
            match std::fs::copy(&source, &target) {
                Ok(_) => count += 1,
                Err(e) => tracing::warn!(error = %e, path = %relative_path, "Backup copy failed"),
            }
        }

        tracing::info!(count = count, backup = %backup_path.display(), "Backup completed");
        Ok(())
    }

    fn cleanup_empty_directories(&self, start_dir: &Path) {
        let workspace_root = &self.workspace_path;
        let mut current = start_dir.to_owned();

        loop {
            if !current.starts_with(workspace_root) {
                break;
            }
            if current == *workspace_root {
                break;
            }
            if std::fs::remove_dir(&current).is_err() {
                break;
            }
            match current.parent() {
                Some(parent) => current = parent.to_owned(),
                None => break,
            }
        }
    }
}
